File: explicit-java/src/main/kotlin/explicit/token/Atomic.kt
package explicit.token

import explicit.api.IToken
import explicit.api.Navigator
import explicit.evaluator.ExplicitRuleEvaluation
import explicit.parser.ExplicitRules
import java.util.*

/**
 * Token implementation which represents a group of tokens which build an atomic unit.
 */
data class Atomic(val tokens: List<IToken>): IToken {
    override fun toString() = "(${tokens.map(IToken::asText).joinToString(" ")})"

    override fun evaluate(rulez: ExplicitRules, navigator: Navigator<String>): ExplicitRuleEvaluation {
        val iterator = tokens.iterator()
        while (iterator.hasNext()) {
            val result = iterator.next().evaluate(rulez, navigator)
            if (!result.result) {
                return ExplicitRuleEvaluation(false, HashMap(), listOf(navigator.getIndex() - 1))
            }
            if (iterator.hasNext()) {
                navigator.next()
            }
        }
        return ExplicitRuleEvaluation(true, HashMap(), listOf(navigator.getIndex() - 1))
    }
}

File: explicit-python/explicit_nlu/token/Atomic.py
import logging
from dataclasses import dataclass
from typing import List

from explicit_nlu.api import IToken, Navigator
from explicit_nlu.evaluator import ExplicitEvaluation
from explicit_nlu.parser import ExplicitRules


#
# Token implementation which represents a group of tokens which build an atomic unit.
#
@dataclass
class Atomic(IToken):
    tokens: List[IToken]

    def __str__(self):
        return " ".join(list(map(IToken.as_text, self.tokens)))

    def __repr__(self):
        return self.__str__()

    def evaluate(self, rulez: ExplicitRules, navigator: Navigator) -> ExplicitEvaluation:
        logging.debug("evaluate atomic")

        for i, entry in enumerate(self.tokens):
            result = entry.evaluate(rulez, navigator)

            if not result.result:
                return ExplicitEvaluation(False, dict(), [navigator.get_absolute_index() - 1])

            if i < len(self.tokens) - 1:
                navigator.__next__()

        return ExplicitEvaluation(True, dict(), [navigator.get_absolute_index() - 1])
